tools/devservices_healthcheck.py

```diff
@@ -1,47 +1,183 @@
-    postgres_healthcheck = ["docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"]
-    kafka_healthcheck = [
-        "docker",
-        "exec",
-        "sentry_kafka",
-        "kafka-topics",
-        "--zookeeper",
-        # TODO: sentry_zookeeper:2181 doesn't work in CI, but 127.0.0.1 doesn't work locally
-        "127.0.0.1:2181",
-        "--list",
-    ]
-
-    healthchecks = [postgres_healthcheck]
-    if os.getenv("NEED_KAFKA") == "true":
-        healthchecks.append(kafka_healthcheck)
-
-    for check in healthchecks:
-        run_cmd(check)
+    checks = []
+    hc = all_service_healthchecks[id]
+    for dep in hc.deps:
+        dep_checks = get_services_to_check(dep)
+        for d in dep_checks:
+            checks.append(d)
+    checks.append(id)
+    return checks
+
+
+def check_health(ids: list[str]) -> None:
+    checks = []
+    for id in ids:
+        s = get_services_to_check(id)
+        checks += s
+
+    # dict.fromkeys is used to remove duplicates while maintaining order
+    for name in dict.fromkeys(checks):
+        print(f"Checking service {name}")
+        hc = all_service_healthchecks[name]
+        print(f"Checking '{hc.container_name}' is running...")
+        ls = " ".join(list(set(checks)))
+        try:
+            run_with_retries(hc.check_container, hc.retries, hc.timeout_secs)
+        except HealthcheckError:
+            raise HealthcheckError(
+                f"Container '{hc.container_name}' is not running.\n"
+                + f"    Start service: sentry devservices up {hc.id}\n"
+                + f"    Restart all services: sentry devservices down {ls} && sentry devservices up {ls}"
+            )
+
+        if hc.check is not None:
+            print(f"Checking '{hc.container_name}' container health...")
+            try:
+                run_with_retries(hc.check, hc.retries, hc.timeout_secs)
+            except HealthcheckError:
+                raise HealthcheckError(
+                    f"Container '{hc.container_name}' does not appear to be healthy.\n"
+                    + f"    Restart service: sentry devservices down {hc.id} && sentry devservices up {hc.id}\n"
+                    + f"    Restart all services: sentry devservices down {ls} && sentry devservices up {ls}"
+                )
+
+
+def main(argv: Sequence[str] | None = None) -> None:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--service",
+        action="append",
+        help="The services you wish to check on. Defaults to all services.",
+    )
+
+    args = parser.parse_args(argv)
+    services = args.service
+
+    healthchecks = services
+    if healthchecks is None:
+        healthchecks = []
+        for k in all_service_healthchecks:
+            if all_service_healthchecks[k].check_by_default:
+                healthchecks.append(k)
+
+    try:
+        check_health(healthchecks)
+    except HealthcheckError as e:
+        print(e)
+        raise SystemExit(1)
```

Comment:
this can be `raise SystemExit(e)` -- it'll print the stringified version to stderr and exit nonzero

if mypy complains it might need to be `raise SystemExit(str(e))` but either way